### PR TITLE
Feature/0712 model base class

### PR DESCRIPTION
**Pull request by bob-carpenter**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Update the basic command (`command.hpp`) to instantiate the services with a base model class (`stan::model::base_model`) and build a reference using a new function defined in the generated model code as of [stan-dev/stan PR 2785](https://github.com/stan-dev/stan/pull/2785).

#### Intended Effect:

Allow the command to be compiled independently of the model to speed up compilation times.

#### How to Verify:

All functionality should still work as is.  

#### Side Effects:

User-facing behavior is unchanged.  

There may be performance issues on some platforms as we've seen one model in our test suite take longer on our old Mac pro but not on other platforms.

#### Documentation:

There's no user-facing documentation and both code and makefile changes are minimal.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
These tests are going to fail until stan-dev/stan PR 2785](https://github.com/stan-dev/stan/pull/2785) is merged.

**Comment by bob-carpenter** [comment]
Sorry, this branch isn't even working on my machine.  I'll ping this again when it's working.

**Comment by bob-carpenter** [comment]
This is ready to review now.

**Comment by wds15** [comment]
As pointed out on discourse:

"new with this PR (to me) is that now CmdStan picks up stuff which are defined in the Stan makefiles - at least this is how we should handle it. Right now the additional make targets are defined in both repos - which is not good, I think."

**Comment by mitzimorris** [comment]
>  Right now the additional make targets are defined in both repos - which is not good

pre-existing redundancies between the CmdStan and Stan makefiles should be addressed via a different issue.

in this PR, the files `makefile` and `make/program` were changed.   otherwise, minimal changes have been made to either CmdStan or Stan makefile in the past 10 months.

**Comment by wds15** [comment]
Sure, pre-existing redundancies are not part of this PR - it's just that I have the impression that you do introduce additional redundancies with this PR. If that's right then you create technical debt and that should not be the case. Let me have a look into it again and maybe I can suggest something here.

Anyway, right now I am getting this error:

```
make: *** No rule to make target `stan/src/stan/model/model_header.d', needed by `build'.  Stop.
make: *** Waiting for unfinished jobs....
```

I have cmdstan at this branch and the stan repository is moved to the branch `feature/2759-model-base`.  Is something wrong with what I have done?

File: make/program [hunk 1]
```diff
@@ -2,6 +2,7 @@
 # Models (to be passed through stanc)
 ##
 CMDSTAN_MAIN ?= src/cmdstan/main.cpp
+CMDSTAN_MAIN_O ?= src/cmdstan/main.o
```
Comment (wds15) [inline make/program]: Please use the patsubst command to derive from CMDSTAN_MAIN the value for CMDSTAN_MAIN_O.

File: make/program [hunk 2]
```diff
@@ -23,22 +24,26 @@ ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
 $(STAN_TARGETS) examples/bernoulli/bernoulli$(EXE) $(patsubst %.stan,%$(EXE),$(wildcard src/test/test-models/*.stan)) : CXXFLAGS_PROGRAM += -include $(USER_HEADER)
 endif
 
-
 %.hpp : %.stan bin/stanc$(EXE)
 	@echo ''
 	@echo '--- Translating Stan model to C++ code ---'
 	$(WINE) bin/stanc$(EXE) $(STANCFLAGS) --o=$@ $<
 
+%.d: %.hpp
```
Comment (wds15) [inline make/program]: This definition here overrides the one included in the top makefile which pulls in math/make/dependencies. Could you please document that you do this here on purpose. The alternative is to not include the dependencies from math which we may not need actually for cmdstan?

File: make/program [hunk 3]
```diff
@@ -23,22 +24,26 @@ ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
...
 
 .PRECIOUS: %.hpp
-%$(EXE) : %.hpp $(CMDSTAN_MAIN) $(LIBSUNDIALS) $(MPI_TARGETS)
+%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS)
 	@echo ''
-	@echo '--- Linking C++ model ---'
-	$(LINK.cpp) $(CXXFLAGS_PROGRAM) -include $< $(CMDSTAN_MAIN) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(OUTPUT_OPTION)
+	@echo '--- Compiling, linking C++ code ---'
+	cp $*.hpp $*.cpp; \
+    $(LINK.cpp) $(CXXFLAGS_PROGRAM) -c -o $*.o $*.cpp ; \
```
Comment (wds15) [inline make/program]: Oh.. this is less ideal, I think. You can make the compiler interpret the hpp file just like a c++ file with the `-x c++` option. So I would suggest to replace 38-41 with this:

```
        $(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $*.o $<
        $(LINK.cpp) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(CMDSTAN_MAIN_O) $*.o $(OUTPUT_OPTION)
```

This way you pickup the right things for compilation (first line) and for linking (second line). Moreover, you do not need to do this copying and removing business.

File: makefile [hunk 4]
```diff
@@ -151,7 +151,7 @@ build-mpi: $(MPI_TARGETS)
 	@echo '--- boost mpi bindings built ---'
 
 .PHONY: build
-build: bin/stanc$(EXE) bin/stansummary$(EXE) bin/print$(EXE) bin/diagnose$(EXE) $(LIBSUNDIALS) $(MPI_TARGETS)
+build: bin/stanc$(EXE) bin/stansummary$(EXE) bin/print$(EXE) bin/diagnose$(EXE) $(LIBSUNDIALS) $(MPI_TARGETS) $(CMDSTAN_MAIN_O) $(STAN)src/stan/model/model_header.d
```
Comment (wds15) [inline makefile]: The `model_header.d` should just be removed here, I think. At least this caused errors for me and I don't quite see the need for this.

File: makefile [hunk 5]
```diff
@@ -182,6 +182,7 @@ clean-manual:
 
 clean-all: clean clean-deps clean-libraries clean-manual
 	$(RM) -r bin
+	$(RM) -r src/cmdstan/main.o
```
Comment (wds15) [inline makefile]: You want to delete `$(CMDSTAN_MAIN_O)` here and not directly name the file defined in the variable.

File: make/program (hunk 2 above)
Comment (mitzimorris) [reply make/program]: removing the override.
we need the dependencies rule.

Comment (wds15) [reply make/program]: Ok...once you do that, then the model_header.o dependency can be build. The only downside is that then you will build the dependency thing for the stan model hpp file... which wastes 1s of compile time as it anyway gets recreated with any model update...but ok; still a win.

File: makefile (hunk 4 above)
Comment (wds15) [reply makefile]: this can stay now once you remove the %.d override, but please verify.

**Comment by mitzimorris** [comment]
@wds15 I've addressed your comments - many thanks!

the makefile rule for the dependency file (.d) gets triggered whenever a model is compiled - this adds about 10% overhead to model compilation.

I agree that there is a lot of overlap between cmdstan and stan makefiles that could/should be reduced.  if the monorepo happens, it will solve this problem.  I've reached out to @syclik for further help with the makefiles - I'm always glad to work on reducing our technical debt.

**Comment by syclik** [comment]
There's some sort of weird makefile circular dependency thing happening where I can't run this:
```
./runCmdStanTests.py src/test/interface/metric_test.cpp
```

It just gets into an infinite loop. I'm trying to track down what it is.

**Comment by syclik** [comment]
There's definitely some linker error. I wasn't able to track down what's causing it. I'll try to look again tomorrow.

**Comment by syclik** [comment]
Ok. Found it.

There are a few errors happening all at once. We want to be building the `.o` file from a Stan file. Then linking the `.o` into the test executable, not including it from the source.

I think we can make that happen. I'm just crashing now and will have to pick it up tomorrow.

**Comment by wds15** [comment]
Hmm... when I do the interface tests I am getting a different error about a missing file:

```
[08:37:21][cmdstan]$ ./runCmdStanTests.py src/test/interface/metric_test.cpp
------------------------------------------------------------
make build

--- CmdStan v2.19.1 built ---
------------------------------------------------------------
make test/interface/metric_test
/opt/rh/devtoolset-6/root/usr/bin/g++ -std=c++1y -pthread -Wno-sign-compare     -O3 -I src -I stan/src -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.69.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include -I stan/lib/stan_math/lib/gtest_1.8.1/include -I stan/lib/stan_math/lib/gtest_1.8.1 -I stan/lib/stan_math/lib/gtest_1.8.1/include -I stan/lib/stan_math/lib/gtest_1.8.1      -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_USE_OWN_TR1_TUPLE  -c src/test/interface/metric_test.cpp -o test/interface/metric_test.o
src/test/interface/metric_test.cpp:6:39: fatal error: test/test-models/proper.hpp: No such file or directory
 #include <test/test-models/proper.hpp>
                                       ^
compilation terminated.
make: *** [test/interface/metric_test.o] Error 1
make test/interface/metric_test failed
exit now (07/16/19 08:37:34 CEST)
```

And I cannot find the `proper.hpp` in my tree. Is that maybe still on your local disk, @mitzimorris ?

**Comment by mitzimorris** [comment]
@wds15, @syclik - agreed - still problems with the unit tests and/or makefiles for them.
working on it.

**Comment by syclik** [comment]
@wds15, `proper.hpp` should be built on the fly: https://github.com/stan-dev/cmdstan/blob/develop/make/tests#L60

It's in the test target there. That should create the executable and that should create the `.hpp` file too. There's a chance that make is eagerly removing the temporary target. That might be why you don't have it.

**Comment by mitzimorris** [comment]
I had to tweak the `make/tests` makefile along with the Python test runner `runCmdStanTests.py` so that:

-  test model .hpp files exist before make is called on unit test programs
- when makefile is used to compile Stan model, make doesn't generate a `.d` file for the model.  this shaves about 10% off the make time.   for example model `example/bernoulli/bernoulli.stan`:

make from `.stan` -> compiled model:
```
~/github/stan-dev/cmdstan (feature/0712-model-base-class)> time make examples/bernoulli/bernoulli

--- Translating Stan model to C++ code ---
bin/stanc  --o=examples/bernoulli/bernoulli.hpp examples/bernoulli/bernoulli.stan
Model name=bernoulli_model
Input file=examples/bernoulli/bernoulli.stan
Output file=examples/bernoulli/bernoulli.hpp

--- Compiling, linking C++ code ---
clang++ -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O3 -I src -I stan/src -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.69.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include    -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION     -c -include-pch stan/src/stan/model/model_header.hpp.gch -x c++ -o examples/bernoulli/bernoulli.o examples/bernoulli/bernoulli.hpp
clang++ -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O3 -I src -I stan/src -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.69.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include    -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION                   stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_nvecserial.a stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_cvodes.a stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_idas.a  src/cmdstan/main.o examples/bernoulli/bernoulli.o -o examples/bernoulli/bernoulli

real	0m6.743s
user	0m6.389s
sys	0m0.329s
```

File: makefile (hunk 5 above)
Comment (wds15) [reply makefile]: Can this be adjusted?

**Comment by mitzimorris** [comment]
@wds15 - done

w/r/t getting your head around script runCmdStanTests.py - and it's real headache of a script -
what's going on is that the script is invoked using the full path name to the unit test files -
e.g. `./runCmdStanTests.py src/test/interface/command_test.cpp` 

then the helper function `mungeName` strips off the leading `src/` and the trailing `.cpp` to run command:  `make test/interface/command_test`

for targets `test/interface/...` the rules in the file `make/tests` come into play - (cf `make/test` lines 5-13)

because `src/cmdstan/command.hpp` has a declaration but not definition of the model base class method `new_model` - https://github.com/stan-dev/cmdstan/blob/d403f0c6a368f6be672228f5bdec65099d06edd2/src/cmdstan/command.hpp#L54-L57

the unit test files that need to run a command (`command_test.hpp` and `metric_test.hpp`) now include  the minimal good test model `src/test/test-models/proper.hpp`.   I added a function to `runCmdStanTests.py` to insure that the test model`.hpp` files are created before the `_test.cpp` files are compiled in order to avoid circular makefile dependencies.

it's all a huge hack.  sorry to put this on you, hope this helps.

**Comment by mitzimorris** [comment]
Stan PR has been merged - restarted last build but fails look like CmdStan tests on Jenkins and Travis aren't seeing Stan develop - what needs to be done to restart tests properly?  @syclik or @serban-nicusor-toptal  or anyone else?

**Comment by wds15** [comment]
You need to merge develop maybe? That will update the submodule def in this PR to the right one.

**Comment by seantalts** [comment]
Latest failure suggests at first glance that maybe this broke how we build MPI?

**Comment by mitzimorris** [comment]
@wds15 - MPI failure - any insights?

**Comment by wds15** [comment]
I can take a look at the MPI trouble...don't see anything obvious instantly.

**Comment by mitzimorris** [comment]
@syclik - MPI tests are failing - could you check the makefile changes - I can't see what's missing.

**Comment by mitzimorris** [comment]
@syclik or @wds15 - most definitely failing unit tests on linux with mpi - here's the end of the error log:

```
src/cmdstan/main.o: In function `stan::math::mpi_cluster::~mpi_cluster()':
main.cpp:(.text._ZN4stan4math11mpi_clusterD2Ev[_ZN4stan4math11mpi_clusterD5Ev]+0x44): undefined reference to `boost::mpi::environment::~environment()'
collect2: error: ld returned 1 exit status
make/program:38: recipe for target 'src/test/test-models/print_uninitialized' failed
```

what needs to be included and/or linked where?

**Comment by mitzimorris** [comment]
this appears to be the command that's failing - complaints are about
`stan/lib/stan_math/stan/math/prim/arr/functor/mpi_cluster_inst.o`
???

```
mpicxx.openmpi -std=c++1y  -Wno-sign-compare    -Wno-delete-non-virtual-dtor -O3 -I src -I stan/src -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.69.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include -I stan/lib/stan_math/lib/gtest_1.8.1/include -I stan/lib/stan_math/lib/gtest_1.8.1     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION   -DSTAN_MPI -DGTEST_USE_OWN_TR1_TUPLE       -Wl,-L,"/mnt/nvme1n1/workspace/CmdStan_PR-713/stan/lib/stan_math/lib/boost_1.69.0/stage/lib" -Wl,-rpath,"/mnt/nvme1n1/workspace/CmdStan_PR-713/stan/lib/stan_math/lib/boost_1.69.0/stage/lib"         stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_nvecserial.a stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_cvodes.a stan/lib/stan_math/lib/sundials_4.1.0/lib/libsundials_idas.a stan/lib/stan_math/lib/boost_1.69.0/stage/lib/libboost_serialization.so stan/lib/stan_math/lib/boost_1.69.0/stage/lib/libboost_mpi.so stan/lib/stan_math/stan/math/prim/arr/functor/mpi_cluster_inst.o src/cmdstan/main.o src/test/test-models/proper.o -o src/test/test-models/proper
```

**Comment by syclik** [comment]
@mitzimorris, I'm sorry -- I don't know. I looked at it, but I wouldn't know where to go for the MPI build stuff; I can read it, but I can't really maintain it. You'll have to get help from @wds15 or someone else that understands exactly how this is being built.

It might be worth checking that GPU also builds independently of this failure.

What should we do about the state of CmdStan and Stan? Should we revert? cc: @bob-carpenter, @seantalts

**Comment by wds15** [comment]
I looked at this yesterday for a moment, but i wasnˋt able to figure it out of what is going on. I can certainly spend some time on this over the weekend, but today or tomorrow is not possible for me to do that.

Generally you need to dynamically link against the boost stuff and you need to statically link in the mpi cluster inst thing...not sure what is going wrong right now.

**Comment by seantalts** [comment]
Re: reverting Stan - from what I understand, those changes are safe to make without this branch being merged in. The tests seem to be passing in develop too to second that idea.

File: make/program [hunk 6]
```diff
@@ -23,22 +27,22 @@ ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
...
-	@echo '--- Linking C++ model ---'
-	$(LINK.cpp) $(CXXFLAGS_PROGRAM) -include $< $(CMDSTAN_MAIN) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(OUTPUT_OPTION)
+	@echo '--- Compiling, linking C++ code ---'
+	$(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $*.o $<
+	$(LINK.cpp) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(CMDSTAN_MAIN_O) $*.o $(OUTPUT_OPTION)
 
 
 ##
-# Dependencies file
+# Dependencies file for precompiled header
```
Comment (seantalts) [inline make/program]: I don't think this comment is right - these are dependency files for everything. Line 53 includes the one from the model_header file.

**Comment by wds15** [comment]
+1

**Comment by mitzimorris** [comment]
@wds15 - got the makefiles working (thanks @seantalts! you were right about link order) -
wanna do another review and/or sign off?

**Comment by wds15** [comment]
I think all is good... with the latest changes the Jenkins tests are not anymore run. Is that on purpose? I think they should still run.

(This is major hack to ensure things are in place)

**Comment by mitzimorris** [comment]
that's really weird and certainly not intended.

**Comment by seantalts** [comment]
🎉 So glad this made it in! Down to the wire, haha.

**Pull request merged**